### PR TITLE
add volcano gang-schedule integration and e2e test

### LINKED ISSUE #568: Add integration and e2e test for the volcano integration
Both integration and e2e tests for the volcano don't exist, although the mpi-operator has a volcano integration feature.
So we should add integration and e2e test for the volcano, similar to scheduler-plugins.

We are looking for contributors that are familiar with volcano.

For scheduler-plugins, we have the following tests:

- integration test
https://github.com/kubeflow/mpi-operator/blob/21f326d1d2e5f1beed667da0cc61e1d07158ef0c/test/integration/mpi_job_controller_test.go#L466-L474

- e2e test
https://github.com/kubeflow/mpi-operator/blob/21f326d1d2e5f1beed667da0cc61e1d07158ef0c/test/e2e/mpi_job_test.go#L306-L309

/help

### PR DESCRIPTION
**Pull request by lowang-bh**
Fix #568 
commit-1: add integration test for volcano
commit-2: add e2e test for volcano

### CODE REVIEW

**Review by tenzen-y — COMMENTED** [review]
@lowang-bh Thanks for working on this!
Will you work on e2e in this PR? If not, can you update the PR description?

before: `Fix https://github.com/kubeflow/mpi-operator/issues/568 :add integration test for volcano`

after: `part-of: https://github.com/kubeflow/mpi-operator/issues/568 :add integration test for volcano`

File: Makefile [hunk 1]
```diff
@@ -192,3 +193,12 @@ scheduler-plugins-chart: scheduler-plugins-crd
...
 	chmod -R 760 $(PROJECT_DIR)/dep-manifests/scheduler-plugins
+
+.PHONY: volcano-scheduler
+volcano-scheduler:
+	-@GOPATH=/tmp go install volcano.sh/volcano/cmd/scheduler@$(VOLCANO_SCHEDULER_VERSION)
+
+.PHONY: volcano-scheduler-crd
+volcano-scheduler-crd: volcano-scheduler
+	mkdir -p $(PROJECT_DIR)/dep-crds/volcano-scheduler/
+	cp -f /tmp/pkg/mod/volcano.sh/volcano@$(VOLCANO_SCHEDULER_VERSION)/config/crd/bases/* $(PROJECT_DIR)/dep-crds/volcano-scheduler
```
Comment (tenzen-y) [inline Makefile:204]: Using a cached repository would be good like this:

```suggestion
VOLCANO_ROOT = $(shell go list -m -f "{{.Dir}}" volcano.sh/volcano)
.PHONY: volcano-scheduler-crd
volcano-scheduler-crd:
	mkdir -p $(shell pwd)/dep-crds/volcano-scheduler/
	cp -f $(VOLCANO_ROOT)/config/crd/bases/*  $(shell pwd)/dep-crds/volcano-scheduler/
```

Comment (lowang-bh) [reply Makefile:204]: I copied it from `make scheduler-plugins-crd` command. Should we keep same style or change `make scheduler-plugins-crd` also?

File: test/integration/mpi_job_controller_test.go [hunk 2]
```diff
@@ -575,6 +575,120 @@ func TestMPIJobWithSchedulerPlugins(t *testing.T) {
...
+			return false, err
+		}
+		if pg == nil {
+			return false, nil
+		}
+		return pg.Spec.MinMember == updatedMinAvaiable, nil
+	}); err != nil {
+		t.Errorf("Failed updating volcano PodGroup: %v", err)
+	}
+}
```
[the remark this answers is not in the document]
Comment (lowang-bh) [reply test/integration/mpi_job_controller_test.go:692]: How about the latest?

File: Makefile (hunk 1 above)
Comment (tenzen-y) [reply Makefile:204]: If it works for scheduler-plugins, I prefer to apply the same way to the scheduler-plugins.

File: Makefile [hunk 3]
```diff
@@ -192,3 +193,17 @@ scheduler-plugins-chart: scheduler-plugins-crd
...
+
+.PHONY: volcano-scheduler-deploy
+volcano-scheduler-deploy: volcano-scheduler-crd
+	mkdir -p $(PROJECT_DIR)/dep-manifests/volcano-scheduler/
+	cp -f /tmp/pkg/mod/volcano.sh/volcano@$(VOLCANO_SCHEDULER_VERSION)/installer/volcano-development.yaml $(PROJECT_DIR)/dep-manifests/volcano-scheduler/
```
Comment (tenzen-y) [inline Makefile:209]: Add an empty line.

File: test/integration/mpi_job_controller_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply test/integration/mpi_job_controller_test.go:692]: Is there how to verify whether all pods are enqueued to volcanoQueue, except by checking PodGroup resource?

File: Makefile (hunk 3 above)
Comment (lowang-bh) [reply Makefile]: Done

File: test/integration/mpi_job_controller_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lowang-bh) [reply test/integration/mpi_job_controller_test.go:692]: Queue's status will has the allocate resource which sum up all pods in queue.

[the remark this answers is not in the document]
Comment (tenzen-y) [reply test/integration/mpi_job_controller_test.go:692]: I see. Thanks.

File: test/e2e/mpi_job_test.go [hunk 4]
```diff
@@ -425,6 +425,131 @@ var _ = ginkgo.Describe("MPIJob", func() {
...
+		var (
+			ctx                    = context.Background()
+			unschedulableResources = &corev1.ResourceList{
+				corev1.ResourceCPU:    resource.MustParse("100000"),   // unschedulable
+				corev1.ResourceMemory: resource.MustParse("100000Gi"), // unschedulable
+			}
+		)
+
+		ginkgo.BeforeEach(func() {
+			// Set up the scheduler-plugins.
```
Comment (tenzen-y) [inline test/e2e/mpi_job_test.go:441]:
```suggestion
			// Set up the volcano-scheduler.
```

File: Makefile [hunk 5]
```diff
@@ -192,3 +193,17 @@ scheduler-plugins-chart: scheduler-plugins-crd
...
 	chmod -R 760 $(PROJECT_DIR)/dep-manifests/scheduler-plugins
+
+.PHONY: volcano-scheduler
+volcano-scheduler:
+	-@GOPATH=/tmp go install volcano.sh/volcano/cmd/scheduler@$(VOLCANO_SCHEDULER_VERSION)
+
+.PHONY: volcano-scheduler-crd
+volcano-scheduler-crd: volcano-scheduler
+	mkdir -p $(PROJECT_DIR)/dep-crds/volcano-scheduler/
+	cp -f /tmp/pkg/mod/volcano.sh/volcano@$(VOLCANO_SCHEDULER_VERSION)/config/crd/bases/* $(PROJECT_DIR)/dep-crds/volcano-scheduler
```
Comment (tenzen-y) [inline Makefile:204]:
```suggestion
VOLCANO_ROOT = $(shell go list -m -f "{{.Dir}}" volcano.sh/volcano)
.PHONY: volcano-scheduler-crd
volcano-scheduler-crd:
	mkdir -p $(shell pwd)/dep-crds/volcano-scheduler/
	cp -f $(VOLCANO_ROOT)/config/crd/bases/*  $(shell pwd)/dep-crds/volcano-scheduler/
```

Did this approach not work?

**Comment by tenzen-y** [comment]
Overall LGTM. Some nits.

File: Makefile (hunk 5 above)
Comment (lowang-bh) [reply Makefile:204]: we use `volcano.sh/apis` in go.mod for its clientsets api, not `volcano.sh/volcano`. But e2e test need to install yamls provided in `volcano.sh/volcano`

Comment (tenzen-y) [reply Makefile:204]: Ah, I see. That makes sense.
Thanks!

**Comment by tenzen-y** [comment]
Overall LGTM.

@lowang-bh Can you resolve conflicts?

**Review by tenzen-y — COMMENTED** [review]
/lgtm
/assign @alculquicondor

File: test/e2e/mpi_job_test.go [hunk 6]
```diff
@@ -425,6 +425,131 @@ var _ = ginkgo.Describe("MPIJob", func() {
...
+				corev1.ResourceCPU:    resource.MustParse("100000"),   // unschedulable
+				corev1.ResourceMemory: resource.MustParse("100000Gi"), // unschedulable
+			}
+		)
+
+		ginkgo.BeforeEach(func() {
+			// Set up the volcano-scheduler.
+			setupVolcanoScheduler()
+			// Set up the mpi-operator so that the volcano scheduler is used as gang-scheduler.
+			ginkgo.By("Scale-In the deployment to 0")
```
Comment (alculquicondor) [inline test/e2e/mpi_job_test.go:444]: could we make this a function that can be shared with scheduler-plugins setup?

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**